### PR TITLE
i#5036 A64 scatter/gather, part 3: Expand vector+immediate

### PR DESCRIPTION
**Pull request by jackgallagher-arm**
Adds support to drx_expand_scatter_gather() for SVE vector+immediate gather load and scatter store instructions, along with tests.

Issue: #5036

### CODE REVIEW

File: suite/tests/client-interface/drx-scattergather-aarch64.cpp [hunk 1]
```diff
@@ -1889,8 +1923,536 @@ test_st1_scalar_plus_vector()
...
+        unsigned governing_p;
+        unsigned base_z;
+    } registers_used;
+
+    element_size_t stored_value_size;
+
+    expected_values_t expected_values;
+
+    vector_plus_immediate_store_test_case_t(std::string name_, test_func_t func_,
+                                            registers_used_t registers_used_,
```
Comment (abhinav92003) [inline suite/tests/client-interface/drx-scattergather-aarch64.cpp:2241]: Re these identifiers: usually we append a `_` for class/struct members, not formal arguments.

File: clients/drcachesim/tests/scattergather-aarch64.templatex [hunk 2]
```diff
@@ -1,102 +1,129 @@
...
-st1w 64bit unscaled offset: PASS
-st1w 64bit unscaled offset \(repeated offset\): PASS
-st1d 32bit unpacked scaled offset uxtw: PASS
-st1d 32bit unpacked scaled offset sxtw: PASS
-st1d 32bit unpacked unscaled offset uxtw: PASS
-st1d 32bit unpacked unscaled offset sxtw: PASS
-st1d 64bit scaled offset: PASS
-st1d 64bit unscaled offset: PASS
-st1d 64bit unscaled offset \(repeated offset\): PASS
+ld1b scalar\+vector 32bit unscaled offset uxtw: PASS
```
Comment (abhinav92003) [inline clients/drcachesim/tests/scattergather-aarch64.templatex:2]: So this test ensures functional correctness of the app under drcachesim instrumentation. Do we have a test that verifies the generated trace?

**Review by abhinav92003 — APPROVED** [review]

File: ext/drx/scatter_gather_aarch64.c [hunk 3]
```diff
@@ -366,11 +391,12 @@ expand_scalar_plus_vector(void *drcontext, instrlist_t *bb, instr_t *sg_instr,
...
-        if (sg_info->index_reg == sg_info->gather_dst_reg) {
-            /* The dst register is also the index register so we need to preserve the
-             * value of the active elements so we can use them as offsets. We do this by
-             * cpying a 0 value into the dst register using the inverse of the mask_reg as
-             * the governing predicate.
+        if ((sg_info->base_reg == sg_info->gather_dst_reg) ||
+            (sg_info->index_reg == sg_info->gather_dst_reg)) {
+            /* The dst register is also the base/index register so we need to preserve the
+             * value of the active elements so we can use them in the address calculation.
+             * We do this by cpying a 0 value into the dst register using the inverse of
```
[the remark this answers is not in the document]
Comment (jackgallagher-arm) [reply ext/drx/scatter_gather_aarch64.c:398]: I actually did mean "cpying" to indicate it uses the `CPY` instruction. I could change it to `CPYing` if that would make it clearer.

[the remark this answers is not in the document]
Comment (abhinav92003) [reply ext/drx/scatter_gather_aarch64.c:398]: Ah ok. Apologies. Either way is fine.

File: clients/drcachesim/tests/scattergather-aarch64.templatex (hunk 2 above)
Comment (jackgallagher-arm) [reply clients/drcachesim/tests/scattergather-aarch64.templatex:2]: How would we go about verifying the trace? Is there an x86 test that does that? I created the AArch64 scatter/gather tests by creating equivalents to the existing x86 scatter/gather tests.

Comment (abhinav92003) [reply clients/drcachesim/tests/scattergather-aarch64.templatex:2]: We have offline-allasm-scattergather-basic-counts.templatex that verifies basic counts.

Comment (jackgallagher-arm) [reply clients/drcachesim/tests/scattergather-aarch64.templatex:2]: Thanks. I'll look at adding an equivalent test in a future PR.

File: suite/tests/client-interface/drx-scattergather-aarch64.cpp (hunk 1 above)
Comment (jackgallagher-arm) [reply suite/tests/client-interface/drx-scattergather-aarch64.cpp]: Done.

**Pull request merged**